assets/app/features/upgrade/steps.factory.js

```diff
@@ -95,5 +98,66 @@
                 }
             ];
         }
+
+        function stepByState(state) {
+            return _.find(factory.steps, function (o) { return o.state === state; });
+        }
+
+        function stepByID(id) {
+            return _.find(factory.steps, function (o) { return o.id === id; });
+        }
+
+        function lastStateForRestore(statusData) {
+            var mapping = {
+                'upgrade_prechecks': 'upgrade-landing',
+                'upgrade_prepare': 'upgrade-landing',
+                'admin_backup': 'upgrade.backup',
+                'admin_repo_checks': 'upgrade.administration-repository-checks',
+                'admin_upgrade': 'upgrade.upgrade-administration-server',
+                'database': 'upgrade.database-configuration',
+                'nodes_repo_checks': 'upgrade.nodes-repositories-checks',
+                'nodes_services': 'upgrade.openstack-services',
+                'nodes_db_dump': 'upgrade.openstack-services',
+                'nodes_upgrade': 'upgrade.upgrade-nodes',
+                'finished': 'TODO(skazi): no UI for this step yet',
+            };
+
+            var currentStep = statusData.current_step,
+                steps = statusData.steps,
+                currentState = mapping[currentStep];
+
+            // select previous step if current is still pending
+            if (steps[currentStep].status === STEP_STATES.pending) {
+                var storedLastState = localStorage.getItem(UPGRADE_LAST_STATE_KEY);
+
+                // if there is some stored state and it's pointing to page for current pending step
+                // show this instead of the previous one, otherwise apply the logic below to find
+                // proper page to show
+                if (storedLastState === currentState) {
+                    return currentState;
+                }
+
+                var currentStepData = stepByState(currentState);
```

Comment:
Why would I do that? It's not C. See: http://stackoverflow.com/questions/5053073/is-defining-every-variable-at-the-top-always-the-best-approach